File: bindings/go/backend/log.go
package backend

import "context"

type LogLevel string

const (
	LogLevelDebug LogLevel = "debug"
	LogLevelInfo  LogLevel = "info"
	LogLevelWarn  LogLevel = "warn"
	LogLevelError LogLevel = "error"
)

type LogEvent struct {
	Name      string
	Level     LogLevel
	EventName string
	Fields    map[string]JSON
}

type LogSignalDefinition struct {
	Name      string
	Level     LogLevel
	EventName string
	Fields    []FieldDescriptor
}

type LogSignalRegistration struct {
	RegisteredNew bool
	Definition    LogSignalDefinition
}

type LogSink interface {
	// RegisterDefinition is idempotent. Re-registering an identical definition is
	// a no-op; incompatible redefinition should return an error.
	RegisterDefinition(ctx context.Context, backend Backend, definition LogSignalDefinition) (LogSignalRegistration, error)

	RegisterDefinitionTx(ctx context.Context, tx Transaction, definition LogSignalDefinition) (LogSignalRegistration, error)

	InspectDefinition(ctx context.Context, backend Backend, name string) (*LogSignalDefinition, error)

	InspectDefinitionTx(ctx context.Context, tx Transaction, name string) (*LogSignalDefinition, error)

	// EmitLogTx stages the event in the caller's OCC transaction. Commit makes it
	// visible to exporters; rollback drops it.
	EmitLog(ctx context.Context, backend Backend, event LogEvent) error

	EmitLogTx(ctx context.Context, tx Transaction, event LogEvent) error
}

File: bindings/rust/log.rs
use std::collections::BTreeMap;

use crate::backend::{Backend, BackendResult, FieldDescriptor};
use crate::json::Json;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum LogLevel {
    Debug,
    Info,
    Warn,
    Error,
}

#[derive(Debug, Clone, PartialEq)]
pub struct LogEvent {
    pub name: String,
    pub level: LogLevel,
    pub event_name: String,
    pub fields: BTreeMap<String, Json>,
}

#[derive(Debug, Clone)]
pub struct LogDefinition {
    pub name: String,
    pub level: LogLevel,
    pub event_name: String,
    pub fields: Vec<FieldDescriptor>,
}

#[derive(Debug, Clone)]
pub struct LogDefinitionRegistration {
    pub registered_new: bool,
    pub definition: LogDefinition,
}

pub trait LogSink<B: Backend> {
    fn register_definition(
        &self,
        backend: &B,
        definition: &LogDefinition,
    ) -> BackendResult<LogDefinitionRegistration>;

    fn register_definition_tx(
        &self,
        tx: &mut B::Tx,
        definition: &LogDefinition,
    ) -> BackendResult<LogDefinitionRegistration>;

    fn inspect_definition(&self, backend: &B, name: &str) -> BackendResult<Option<LogDefinition>>;

    fn inspect_definition_tx(
        &self,
        tx: &mut B::Tx,
        name: &str,
    ) -> BackendResult<Option<LogDefinition>>;

    /// Transactional emits are staged in the caller's OCC transaction. Commit
    /// makes the log visible to exporters; rollback drops it.
    fn emit_log(&self, backend: &B, event: &LogEvent) -> BackendResult<()>;

    fn emit_log_tx(&self, tx: &mut B::Tx, event: &LogEvent) -> BackendResult<()>;
}
